### DISCUSSION TITLE
Spring Boot Connector implementation

### DISCUSSION THREAD

**Discussion by MoritzTheile** [OpusCapita/filemanager#321] [Ideas]
Great work! I would love to use the Client Code with our Spring Boot (Java) Backend. Looking at the traffic a implementation of the backend service doesn't seem to hard. However , it would be a lot easier not to have to reverse engineer the API by inspecting the traffic but to have some kind of API interface specification. Can you help with that? Pointing me to the right parts in the code would help already.
I would also be happy to share the project under same license. It shall consist of the API with different implementations AWS S3, local FS and GoogleDrive.

**Comment by achernyakevich-sc** [comment]
@MoritzTheile In reality development of good implementation of the Java Connector could be not so simple as it could looks like. For sure, implementation of simple connector that do not care about security related things (authorized users only, permissions, etc.) - is not a hard task and depends on Connector API that client part of component expect to work with.

It looks that in next week we could have one developer who will have some free time and wish to experiment in this area. So our plan is:
- get documented _Connector API_ expected by Filemanager client component.
- get implemented simple _Filemanager Java Connector_ as a Java library that could be included to the project as Maven dependency.
- think how we could introduce security related this in configurable manner to the _Filemanager Java Connector_.
- get implemented enhanced version of the _Filemanager Java Connector_.

It is a maximum plan. Implementation depends on availability of time that we could spend for it.

We will try to keep community up to date.

**Reply by MoritzTheile** [reply]
Thanks for the reply and sorry for my delay (long holiday...).

In the meantime I managed to implement a working S3 connector by reverse engeneering the http traffic and looking at your [docs](https://github.com/OpusCapita/filemanager/blob/master/packages/server-nodejs/docs/api-overview.md).

The code is mixed into our production system. However, I share some snippets with you. Maybe it helps:

Below is the rest service interface part. If you are interested in the S3 connecting part (folder creation was a challenge) drop me a line.
`
import java.text.SimpleDateFormat;
import java.util.Base64;
import java.util.Date;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.InputStreamResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.lang.Nullable;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.lithodat.app.litho.service.other.FileBrowserService;
import com.lithodat.app.litho.service.other.FileBrowserService.ContentPath;
import com.lithodat.app.litho.service.other.FileBrowserService.DirPath;
import com.lithodat.app.litho.service.other.FileBrowserService.LithoFile;
import com.lithodat.app.litho.web.rest.other.filebrowserdtos.Capabilities;
import com.lithodat.app.litho.web.rest.other.filebrowserdtos.FileStatsResource;
import com.lithodat.app.litho.web.rest.other.filebrowserdtos.FileStatsResourceItems;
import com.lithodat.app.litho.web.rest.other.filebrowserdtos.FileUploadForm;

/**
 * REST controller implementing file browser api.
 * see https://github.com/OpusCapita/filemanager/blob/master/packages/server-nodejs/docs/api-overview.md
 */
@RestController
@RequestMapping("/api/other/browser")
public class OpuscapitaFilemanagerResource {

	@Autowired
	private FileBrowserService fileBrowserService;

	/**
	 * Creating this service was a hard nut to crack, 
	 * because sometimes it consumes form data (when uploading a file) and sometimes
	 * it consumes JSON (when creating a directory).
	 * The @Nullable annotation helps to allow requests without JSON body.
	 */
	@RequestMapping(value = "/{resource}/{resourceId}/files", consumes = "*/*", method = RequestMethod.POST)
	public ResponseEntity<FileStatsResource> postFiles(@Nullable @RequestBody InputStreamResource bodyAsStream, FileUploadForm fileUploadForm) throws Exception {

		if (bodyAsStream != null) {

			FileUploadForm dirJson = new ObjectMapper().readValue(bodyAsStream.getInputStream(), FileUploadForm.class);

			String parentPath = new String(Base64.getDecoder().decode(dirJson.getParentId()));

			LithoFile createContent = fileBrowserService.createDir(new DirPath(parentPath + dirJson.getName() + "/"));

			return ResponseEntity.ok().body(lithoRoot2FileStatsResource(createContent));

		}

		if (fileUploadForm != null) {

			String parentPath = new String(Base64.getDecoder().decode(fileUploadForm.getParentId()));

			LithoFile createContent = fileBrowserService.createContent(new ContentPath(parentPath + fileUploadForm.getFiles().getOriginalFilename()), fileUploadForm.getFiles().getBytes());

			return ResponseEntity.ok().body(lithoRoot2FileStatsResource(createContent));

		}

		return ResponseEntity.noContent().build();

	}

	@GetMapping("/{resource}/{resourceId}/files")
	public ResponseEntity<FileStatsResource> getFiles(@PathVariable String resource, @PathVariable String resourceId) throws Exception {

		LithoFile dir = fileBrowserService.createDir(new DirPath(resource + "_" + resourceId + "/"));

		return ResponseEntity.ok().body(lithoRoot2FileStatsResource(dir));

	}

	@GetMapping("/{resource}/{resourceId}/files/{fileId}")
	public ResponseEntity<FileStatsResource> getFile(@PathVariable String fileId) throws Exception {

		fileId = new String(Base64.getDecoder().decode(fileId));

		LithoFile lithoFile = fileBrowserService.getFile(fileId);

		FileStatsResource fileStatsResource = lithoFile2FileStatsResource(lithoFile);

		return ResponseEntity.ok().body(fileStatsResource);

	}

	@DeleteMapping("/{resource}/{resourceId}/files/{fileId}")
	public ResponseEntity<Void> deleteFiles(@PathVariable String fileId) throws Exception {

		fileId = new String(Base64.getDecoder().decode(fileId));

		if (fileId.endsWith("/")) {

			fileBrowserService.deleteDir(new DirPath(fileId));

		} else {

			fileBrowserService.deleteContent(new ContentPath(fileId));

		}

		return ResponseEntity.noContent().build();
	}

	@GetMapping("/{resource}/{resourceId}/files/{fileId}/children")
	public ResponseEntity<FileStatsResourceItems> getChildren(@PathVariable String fileId, @RequestParam String orderBy, @RequestParam String orderDirection) throws Exception {

		fileId = new String(Base64.getDecoder().decode(fileId));

		List<LithoFile> listFiles = fileBrowserService.listDir(new DirPath(fileId));

		FileStatsResourceItems fileStatsResourceItems = new FileStatsResourceItems();

		for (LithoFile lithoFile : listFiles) {

			fileStatsResourceItems.getItems().add(lithoFile2FileStatsResource(lithoFile));

		}

		return ResponseEntity.ok().body(fileStatsResourceItems);

	}

	private FileStatsResource lithoRoot2FileStatsResource(LithoFile rootDir) {

		FileStatsResource fileStatsResource = new FileStatsResource();

		fileStatsResource.setId(Base64.getEncoder().encodeToString(rootDir.path.getBytes()));
		fileStatsResource.setName(rootDir.name);

		{
			Capabilities capabilities = new Capabilities();
			fileStatsResource.setCapabilities(capabilities);
			capabilities.setCanDelete(false);
			capabilities.setCanRename(false);
			capabilities.setCanCopy(false);
			capabilities.setCanEdit(false);
			capabilities.setCanDownload(false);
			capabilities.setCanListChildren(true);
			capabilities.setCanAddChildren(true);
			capabilities.setCanRemoveChildren(true);
		}

		fileStatsResource.setType("dir");

		return fileStatsResource;
	}

	private FileStatsResource lithoFile2FileStatsResource(LithoFile lithoFile) {

		FileStatsResource fileStatsResource = new FileStatsResource();

		fileStatsResource.setId(Base64.getEncoder().encodeToString(lithoFile.path.getBytes()));
		fileStatsResource.setName(lithoFile.name);
		fileStatsResource.setCreatedTime(toISO8601(lithoFile.lastModified));
		fileStatsResource.setModifiedTime(toISO8601(lithoFile.lastModified));

		{
			Capabilities capabilities = new Capabilities();
			fileStatsResource.setCapabilities(capabilities);
			capabilities.setCanDelete(true);
			capabilities.setCanRename(true);
			capabilities.setCanCopy(false);
			capabilities.setCanEdit(false);
			capabilities.setCanDownload(true);

			if (lithoFile.isDir) {

				capabilities.setCanListChildren(true);
				capabilities.setCanAddChildren(true);
				capabilities.setCanRemoveChildren(true);

			}
		}

		if (lithoFile.isDir) {

			fileStatsResource.setType("dir");

		} else {

			fileStatsResource.setType("file");

			fileStatsResource.setSize(lithoFile.contentLength + "");

		}

		fileStatsResource.setParentId(getParentPath(lithoFile.path));

		// Ancestors don't seem to be necessary for functioning...
		//		   "ancestors":[
		//		      
		//		   ]

		return fileStatsResource;

	}

	private static SimpleDateFormat sdf = new SimpleDateFormat(
			"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

	private String toISO8601(Date date) {
		return sdf.format(date);
	}

	private String getParentPath(String stringPath) {

		String[] pathAsArray = stringPath.split("/");

		String parentPath = "";

		for (int i = 0; i < pathAsArray.length - 1; i++) {
			parentPath += pathAsArray[i] + "/";
		}

		return parentPath;

	}

	@GetMapping("/{resource}/{resourceId}/files/{fileId}/search")
	public ResponseEntity<FileStatsResource> searchFiles(@PathVariable String fileId, @RequestParam String itemNameSubstring, @RequestParam String itemNameCaseSensitive, @RequestParam String itemType, @RequestParam String recursive, @RequestParam String fileContentSubstring, @RequestParam String fileContentCaseSensitive) throws Exception {
		fileId = new String(Base64.getDecoder().decode(fileId));
		throw new Exception("not implemented yet");
	}

	/*
	 * request to rename: {name: "Careem-on-call-blog.jpegs"}
	 */
	@PatchMapping("/{resource}/{resourceId}/files/{fileId}")
	public ResponseEntity<FileStatsResource> patchFile(@PathVariable String fileId) throws Exception {
		fileId = new String(Base64.getDecoder().decode(fileId));
		throw new Exception("not implemented yet");
	}

	@GetMapping("/{resource}/{resourceId}/download")
	public ResponseEntity<Resource> downloadFile(@RequestParam String items) throws Exception {

		items = new String(Base64.getDecoder().decode(items));

		byte[] content = fileBrowserService.getContent(new ContentPath(items));

		ByteArrayResource byteArrayResource = new ByteArrayResource(content);

		LithoFile lithoFile = fileBrowserService.getFile(items);

		HttpHeaders headers = new HttpHeaders();

		headers.add(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=" + lithoFile.name);
		headers.add(HttpHeaders.CONTENT_TYPE, lithoFile.contentType);

		return ResponseEntity.ok()
				.headers(headers)
				.contentLength(byteArrayResource.contentLength())
				.body(byteArrayResource);
	}


-------------------------------------------



import java.util.ArrayList;
import java.util.List;

public class FileStatsResource {
	// id: <string>,
	private String id;
	// name: <string>,
	private String name;
	// type: <string>, // "dir" or "file"
	private String type;
	// createdTime: <string>,
	private String createdTime;
	// modifiedTime: <string>,
	private String modifiedTime;
	// ?size: <string>, // for files only
	private String size;
	// ?parentId: <string>, // for non-root only
	private String parentId;
	// ancestors: [<file stats resource>, ...], // list of ancestors starting with user root dir, empty for the root
	private List<FileStatsResource> ancestors = new ArrayList<>();
	// md5Checksum: <string>, // in v2
	private String md5Checksum;
	// capabilities: <Capabilities>
	private Capabilities capabilities;

	public String getId() {
		return id;
	}

	public void setId(String id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public String getCreatedTime() {
		return createdTime;
	}

	public void setCreatedTime(String createdTime) {
		this.createdTime = createdTime;
	}

	public String getModifiedTime() {
		return modifiedTime;
	}

	public void setModifiedTime(String modifiedTime) {
		this.modifiedTime = modifiedTime;
	}

	public String getSize() {
		return size;
	}

	public void setSize(String size) {
		this.size = size;
	}

	public String getParentId() {
		return parentId;
	}

	public void setParentId(String parentId) {
		this.parentId = parentId;
	}

	public List<FileStatsResource> getAncestors() {
		return ancestors;
	}

	public void setAncestors(List<FileStatsResource> ancestors) {
		this.ancestors = ancestors;
	}

	public String getMd5Checksum() {
		return md5Checksum;
	}

	public void setMd5Checksum(String md5Checksum) {
		this.md5Checksum = md5Checksum;
	}

	public Capabilities getCapabilities() {
		return capabilities;
	}

	public void setCapabilities(Capabilities capabilities) {
		this.capabilities = capabilities;
	}

}

----------------------------------------------------------


import java.util.ArrayList;
import java.util.List;

public class FileStatsResourceItems {

	private List<FileStatsResource> items = new ArrayList<>();

	public List<FileStatsResource> getItems() {
		return items;
	}

	public void setItems(List<FileStatsResource> items) {
		this.items = items;
	}

}

--------------------------------------------------------


import org.springframework.web.multipart.MultipartFile;

public class FileUploadForm {

	private String name;
	private String type;
	private String parentId;
	private MultipartFile files;

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public String getParentId() {
		return parentId;
	}

	public void setParentId(String parentId) {
		this.parentId = parentId;
	}

	public MultipartFile getFiles() {
		return files;
	}

	public void setFiles(MultipartFile files) {
		this.files = files;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

}
`

**Reply by MoritzTheile** [reply]
The code I send in previous message is out of shape. However, you will get an idea of what I did. If you see value in it I'm happy to deliver it in a better shape.
